pytorch_lightning/trainer/connectors/accelerator_connector.py

```diff
@@ -436,12 +435,12 @@ def resolve_training_type_plugin(self, training_type: TrainingTypePlugin) -> Tra
         if hasattr(training_type, 'cluster_environment') and getattr(training_type, 'cluster_environment') is None:
             training_type.cluster_environment = self.select_cluster_environment()
 
-        if hasattr(training_type, 'num_nodes') and getattr(training_type, 'num_nodes') is None:
+        if hasattr(training_type, 'num_nodes') and getattr(training_type, 'num_nodes') != self.num_nodes:
+            # set num_nodes for training_type from trainer setting
```

Comment:
right above here we have another patch for parallel_devices. This is another candidate for a setter because it also requires updating num_processes.

Reply:
why the second check for `getattr(training_type, 'num_nodes') != self.num_nodes` ? 
optimization? to avoid resetting the ranks?

Reply:
I would argue the resetting of ranks is desired, if the default was 1 and then we override the real num_nodes value later, we need to update the ranks, it affects the world size (if we need to compute it).
For this reason I never understood why we are patching these values from outside.

Reply:
> right above here we have another patch for parallel_devices. This is another candidate for a setter because it also requires updating num_processes.

@awaelchli , yeah, we do plan to consolidate `parallel_devices`, for simplicity, that would be a separate PR.

Reply:
> why the second check for `getattr(training_type, 'num_nodes') != self.num_nodes` ?
> optimization? to avoid resetting the ranks?

@ananthsub, not for optimization, for some weird reason, if I reset the same num nodes, calling `set_world_ranks` for `DDPSpawnPlugin` caused some test failures.

Reply:
it is not reproducible always. for example test: 

```tests/callbacks/test_stochastic_weight_avg.py::test_swa_callback_ddp_cpu```

reran for several times,  sometimes, it would complain no such file: `/tmp/pytest-of-shuyingsun/pytest-18/test_swa_callback_ddp_cpu0/lightning_logs/version_0/checkpoints/epoch=3-step=11.ckpt` (have not figured out the root cause, might be related to global rank change somehow)

but on the other hand, as `resolve_training_type_plugin` is called in `self.training_type_plugin`, every time, when we call `self.training_type_plugin` (for example in checking is_distributed, or select precision plugin), we call `resolve_training_type_plugin` is not necessary.

I will introduce `_training_type_plugin_resolved` so that it will only resolve/select once.